File: android/src/main/kotlin/app/picsa/capacitorofflinetransfer/CapacitorOfflineTransferPlugin.kt
package app.picsa.capacitorofflinetransfer

import android.Manifest
import android.net.Uri
import android.os.Build
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.PermissionState
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import com.getcapacitor.annotation.PermissionCallback

@CapacitorPlugin(
    name = "OfflineTransfer",
    permissions = [
        Permission(
            alias = "nearby",
            strings = [
                Manifest.permission.ACCESS_COARSE_LOCATION,
                Manifest.permission.ACCESS_FINE_LOCATION,
                Manifest.permission.BLUETOOTH_SCAN,
                Manifest.permission.BLUETOOTH_ADVERTISE,
                Manifest.permission.BLUETOOTH_CONNECT,
                Manifest.permission.NEARBY_WIFI_DEVICES
            ]
        )
    ]
)

class CapacitorOfflineTransferPlugin : Plugin() {

    private val implementation = CapacitorOfflineTransfer()
    private var sessionStartTime: Long = 0

    @PermissionCallback
    private fun permissionsCallback(call: PluginCall) {
        try {
            val state = getPermissionState("nearby")
            if (state == PermissionState.GRANTED) {
                when (call.methodName) {
                    "startAdvertising" -> {
                        sessionStartTime = System.currentTimeMillis()
                        val displayName = call.getString("displayName")
                        implementation.startAdvertising(displayName, call)
                    }
                    "startDiscovery" -> {
                        sessionStartTime = System.currentTimeMillis()
                        implementation.startDiscovery(call)
                    }
                    "requestPermissions" -> {
                        val result = JSObject()
                        result.put("nearby", "granted")
                        call.resolve(result)
                    }
                    else -> call.reject("Unexpected method '${call.methodName}' in permissions callback.")
                }
            } else {
                if (call.methodName == "requestPermissions") {
                    val result = JSObject()
                    result.put("nearby", state?.toString() ?: "denied")
                    call.resolve(result)
                } else {
                    call.reject("Permissions denied. Please enable in Settings.")
                }
            }
        } catch (e: Exception) {
            call.reject("Permission check failed: ${e.message}")
        }
    }


    /**
     * JS-callable method to check permission status.
     * Returns: { nearby: "granted" | "denied" | "prompt" }
     */
    @PluginMethod
    override fun checkPermissions(call: PluginCall) {
        try {
            val state = getPermissionState("nearby")
            val result = JSObject()
            result.put("nearby", state?.toString() ?: "denied")
            call.resolve(result)
        } catch (e: Exception) {
            call.reject("Failed to check permissions: ${e.message}")
        }
    }

    /**
     * JS-callable method to request permissions explicitly.
     * Use this if you want to request permissions before advertising/discovery.
     */
    @PluginMethod
    override fun requestPermissions(call: PluginCall) {
        try {
            requestPermissionForAlias("nearby", call, "permissionsCallback")
        } catch (e: Exception) {
            call.reject("Failed to request permissions: ${e.message}")
        }
    }

    private fun ensurePermissionsForMethod(call: PluginCall) {
        try {
            val state = getPermissionState("nearby")
            if (state == PermissionState.GRANTED) {
                when (call.methodName) {
                    "startAdvertising" -> {
                        sessionStartTime = System.currentTimeMillis()
                        val displayName = call.getString("displayName")
                        implementation.startAdvertising(displayName, call)
                    }
                    "startDiscovery" -> {
                        sessionStartTime = System.currentTimeMillis()
                        implementation.startDiscovery(call)
                    }
                    else -> {
                        call.reject("Method ${call.methodName} not supported here.")
                    }
                }
            } else {
                requestPermissionForAlias("nearby", call, "permissionsCallback")
            }
        } catch (e: Exception) {
            call.reject("Permission check failed: ${e.message}")
        }
    }

    private fun getFileExtension(filePath: String): String {
        return try {
            val uri = Uri.parse(filePath)
            if (uri.scheme == "content") {
                val mimeType = context.contentResolver.getType(uri)
                mimeTypeToExtension(mimeType)
            } else {
                java.io.File(filePath).extension
            }
        } catch (e: Exception) {
            java.io.File(filePath).extension
        }
    }

    private fun mimeTypeToExtension(mimeType: String?): String {
        if (mimeType == null) return ""
        return when (mimeType.lowercase()) {
            "image/jpeg" -> "jpg"
            "image/png" -> "png"
            "image/gif" -> "gif"
            "image/webp" -> "webp"
            "image/bmp" -> "bmp"
            "image/svg+xml" -> "svg"
            "video/mp4" -> "mp4"
            "video/webm" -> "webm"
            "video/quicktime" -> "mov"
            "video/x-matroska" -> "mkv"
            "video/x-msvideo" -> "avi"
            "audio/mpeg" -> "mp3"
            "audio/wav" -> "wav"
            "audio/ogg" -> "ogg"
            "audio/flac" -> "flac"
            "audio/webm" -> "webm"
            "application/json" -> "json"
            "application/xml" -> "xml"
            "application/pdf" -> "pdf"
            "application/zip" -> "zip"
            "application/gzip" -> "gz"
            "application/x-tar" -> "tar"
            "application/x-rar-compressed" -> "rar"
            "application/msword" -> "doc"
            "application/vnd.openxmlformats-officedocument.wordprocessingml.document" -> "docx"
            "application/vnd.ms-excel" -> "xls"
            "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet" -> "xlsx"
            "application/vnd.ms-powerpoint" -> "ppt"
            "application/vnd.openxmlformats-officedocument.presentationml.presentation" -> "pptx"
            "text/plain" -> "txt"
            "text/html" -> "html"
            "text/css" -> "css"
            "text/javascript" -> "js"
            "application/javascript" -> "js"
            "text/csv" -> "csv"
            else -> ""
        }
    }

    override fun load() {
        implementation.load(context, this)
    }

    fun emit(event: String, data: JSObject) {
        notifyListeners(event, data)
    }

    @PluginMethod
    fun initialize(call: PluginCall) {
        try {
            val serviceId = call.getString("serviceId") ?: return call.reject("serviceId is required")
            implementation.initialize(serviceId)
            call.resolve(JSObject().apply { put("success", true) })
        } catch (e: Exception) {
            call.reject("Initialize failed: ${e.message}")
        }
    }

    @PluginMethod
    fun checkCapabilities(call: PluginCall) {
        try {
            val capabilities = implementation.checkCapabilities()
            val result = JSObject()
            result.put("platform", capabilities.platform)
            result.put("transferMethod", capabilities.transferMethod)
            result.put("supportsNearby", capabilities.supportsNearby)
            result.put("isEmulator", capabilities.isEmulator)
            capabilities.reason?.let { result.put("reason", it) }
            call.resolve(result)
        } catch (e: Exception) {
            call.reject("Check capabilities failed: ${e.message}")
        }
    }

    @PluginMethod
    fun startAdvertising(call: PluginCall) {
        try {
            if (implementation.checkCapabilities().transferMethod == "lan") {
                sessionStartTime = System.currentTimeMillis()
                implementation.startAdvertising(call.getString("displayName"), call)
                return
            }
            ensurePermissionsForMethod(call)
        } catch (e: Exception) {
            call.reject("Start advertising failed: ${e.message}")
        }
    }

    @PluginMethod
    fun stopAdvertising(call: PluginCall) {
        try {
            sessionStartTime = 0
            implementation.stopAdvertising()
            call.resolve(JSObject().apply { put("success", true) })
        } catch (e: Exception) {
            call.reject("Stop advertising failed: ${e.message}")
        }
    }

    @PluginMethod
    fun startDiscovery(call: PluginCall) {
        try {
            if (implementation.checkCapabilities().transferMethod == "lan") {
                sessionStartTime = System.currentTimeMillis()
                implementation.startDiscovery(call)
                return
            }
            ensurePermissionsForMethod(call)
        } catch (e: Exception) {
            call.reject("Start discovery failed: ${e.message}")
        }
    }

    @PluginMethod
    fun stopDiscovery(call: PluginCall) {
        try {
            sessionStartTime = 0
            implementation.stopDiscovery()
            call.resolve(JSObject().apply { put("success", true) })
        } catch (e: Exception) {
            call.reject("Stop discovery failed: ${e.message}")
        }
    }

    @PluginMethod
    fun connect(call: PluginCall) {
        try {
            val endpointId = call.getString("endpointId") ?: return call.reject("endpointId is required")
            val displayName = call.getString("displayName")
            implementation.connect(endpointId, displayName)
            call.resolve(JSObject().apply { put("success", true) })
        } catch (e: Exception) {
            call.reject("Connect failed: ${e.message}")
        }
    }

    @PluginMethod
    fun acceptConnection(call: PluginCall) {
        try {
            val endpointId = call.getString("endpointId") ?: return call.reject("endpointId is required")
            implementation.acceptConnection(endpointId)
            call.resolve(JSObject().apply { put("success", true) })
        } catch (e: Exception) {
            call.reject("Accept connection failed: ${e.message}")
        }
    }

    @PluginMethod
    fun rejectConnection(call: PluginCall) {
        try {
            val endpointId = call.getString("endpointId") ?: return call.reject("endpointId is required")
            implementation.rejectConnection(endpointId)
            call.resolve(JSObject().apply { put("success", true) })
        } catch (e: Exception) {
            call.reject("Reject connection failed: ${e.message}")
        }
    }

    @PluginMethod
    fun disconnectFromEndpoint(call: PluginCall) {
        try {
            val endpointId = call.getString("endpointId") ?: return call.reject("endpointId is required")
            implementation.disconnectFromEndpoint(endpointId)
            call.resolve(JSObject().apply { put("success", true) })
        } catch (e: Exception) {
            call.reject("Disconnect failed: ${e.message}")
        }
    }

    @PluginMethod
    fun disconnect(call: PluginCall) {
        try {
            implementation.disconnect()
            call.resolve(JSObject().apply { put("success", true) })
        } catch (e: Exception) {
            call.reject("Disconnect failed: ${e.message}")
        }
    }

    @PluginMethod
    fun sendMessage(call: PluginCall) {
        try {
            val endpointId = call.getString("endpointId") ?: return call.reject("endpointId is required")
            val data = call.getString("data") ?: return call.reject("data is required")
            implementation.sendMessage(endpointId, data)
            call.resolve(JSObject().apply { put("success", true) })
        } catch (e: Exception) {
            call.reject("Send message failed: ${e.message}")
        }
    }

    @PluginMethod
    fun sendFile(call: PluginCall) {
        try {
            val endpointId = call.getString("endpointId") ?: return call.reject("endpointId is required")
            val filePath = call.getString("filePath") ?: return call.reject("filePath is required")

            var fileName = call.getString("fileName")
            if (fileName.isNullOrBlank()) {
                val timestamp = System.currentTimeMillis()
                val uuid = java.util.UUID.randomUUID().toString().take(8)
                val extension = getFileExtension(filePath)
                fileName = if (extension.isNotEmpty()) "file_${timestamp}_${uuid}.$extension" else "file_${timestamp}_${uuid}"
            }

            val payloadId = implementation.sendFile(endpointId, filePath, fileName)
            call.resolve(JSObject().apply { put("payloadId", payloadId ?: fileName) })
        } catch (e: Exception) {
            call.reject("Send file failed: ${e.message}")
        }
    }

    @PluginMethod
    fun setLogLevel(call: PluginCall) {
        try {
            call.resolve(JSObject().apply { put("success", true) })
        } catch (e: Exception) {
            call.reject("Set log level failed: ${e.message}")
        }
    }

    @PluginMethod
    fun getState(call: PluginCall) {
        try {
            call.resolve(buildStateSnapshot())
        } catch (e: Exception) {
            call.reject("Get state failed: ${e.message}")
        }
    }

    @PluginMethod
    fun syncFromPlugin(call: PluginCall) {
        try {
            call.resolve(buildStateSnapshot())
        } catch (e: Exception) {
            call.reject("Sync from plugin failed: ${e.message}")
        }
    }

    private fun buildStateSnapshot(): JSObject {
        val result = JSObject()
        val endpoints = implementation.getDiscoveredEndpoints()
        val connectedEndpoints = implementation.getConnectedEndpoints()
        result.put("endpoints", endpoints)
        result.put("connectedEndpoints", connectedEndpoints)
        result.put("activeTransfers", JSObject())
        result.put("transferHistory", JSArray())
        result.put("stats", JSObject().put("totalBytesTransferred", 0).put("filesTransferred", 0).put("sessionStart", sessionStartTime).put("currentSpeedBps", 0))
        return result
    }
}

File: ios/Sources/CapacitorOfflineTransferPlugin/CapacitorOfflineTransferPlugin.swift
import Foundation
import Capacitor

@objc(OfflineTransferPlugin)
public class CapacitorOfflineTransferPlugin: CAPPlugin, CAPBridgedPlugin, CapacitorOfflineTransferDelegate {
    public let identifier = "CapacitorOfflineTransferPlugin"
    public let jsName = "OfflineTransfer"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkCapabilities", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startAdvertising", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopAdvertising", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startDiscovery", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopDiscovery", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "connect", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "acceptConnection", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "rejectConnection", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "disconnectFromEndpoint", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "disconnect", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sendMessage", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sendFile", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setLogLevel", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getState", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "syncFromPlugin", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = CapacitorOfflineTransfer()
    private var sessionStartTime: Int = 0

    public override func load() {
        implementation.delegate = self
    }

    @objc func initialize(_ call: CAPPluginCall) {
        do {
            guard let serviceId = call.getString("serviceId") else {
                call.reject("serviceId is required")
                return
            }
            implementation.initialize(serviceId: serviceId)
            call.resolve()
        } catch {
            call.reject("Initialize failed: \(error.localizedDescription)")
        }
    }

    @objc func checkCapabilities(_ call: CAPPluginCall) {
        do {
            let capabilities: [String: Any] = [
                "platform": "ios",
                "transferMethod": "nearby",
                "supportsNearby": true,
                "isEmulator": false
            ]
            call.resolve(capabilities)
        } catch {
            call.reject("Check capabilities failed: \(error.localizedDescription)")
        }
    }

    @objc func checkPermissions(_ call: CAPPluginCall) {
        do {
            let result: [String: Any] = [
                "nearby": "granted"
            ]
            call.resolve(result)
        } catch {
            call.reject("Check permissions failed: \(error.localizedDescription)")
        }
    }

    @objc func requestPermissions(_ call: CAPPluginCall) {
        do {
            let result: [String: Any] = [
                "nearby": "granted"
            ]
            call.resolve(result)
        } catch {
            call.reject("Request permissions failed: \(error.localizedDescription)")
        }
    }

    private func ensurePermissions(call: CAPPluginCall, onGranted: @escaping () -> Void) {
        do {
            onGranted()
        } catch {
            call.reject("Permission check failed: \(error.localizedDescription)")
        }
    }

    @objc func startAdvertising(_ call: CAPPluginCall) {
        do {
            ensurePermissions(call: call) {
                self.sessionStartTime = Int(Date().timeIntervalSince1970 * 1000)
                guard let displayName = call.getString("displayName") else {
                    call.reject("displayName is required")
                    return
                }
                self.implementation.startAdvertising(displayName: displayName)
                call.resolve()
            }
        } catch {
            call.reject("Start advertising failed: \(error.localizedDescription)")
        }
    }

    @objc func stopAdvertising(_ call: CAPPluginCall) {
        do {
            sessionStartTime = 0
            implementation.stopAdvertising()
            call.resolve()
        } catch {
            call.reject("Stop advertising failed: \(error.localizedDescription)")
        }
    }

    @objc func startDiscovery(_ call: CAPPluginCall) {
        do {
            ensurePermissions(call: call) {
                self.sessionStartTime = Int(Date().timeIntervalSince1970 * 1000)
                self.implementation.startDiscovery()
                call.resolve()
            }
        } catch {
            call.reject("Start discovery failed: \(error.localizedDescription)")
        }
    }

    @objc func stopDiscovery(_ call: CAPPluginCall) {
        do {
            sessionStartTime = 0
            implementation.stopDiscovery()
            call.resolve()
        } catch {
            call.reject("Stop discovery failed: \(error.localizedDescription)")
        }
    }

    @objc func connect(_ call: CAPPluginCall) {
        do {
            guard let endpointId = call.getString("endpointId") else {
                call.reject("endpointId is required")
                return
            }
            implementation.connect(endpointId: endpointId)
            call.resolve()
        } catch {
            call.reject("Connect failed: \(error.localizedDescription)")
        }
    }

    @objc func acceptConnection(_ call: CAPPluginCall) {
        do {
            guard let endpointId = call.getString("endpointId") else {
                call.reject("endpointId is required")
                return
            }
            implementation.acceptConnection(endpointId: endpointId)
            call.resolve()
        } catch {
            call.reject("Accept connection failed: \(error.localizedDescription)")
        }
    }

    @objc func rejectConnection(_ call: CAPPluginCall) {
        do {
            guard let endpointId = call.getString("endpointId") else {
                call.reject("endpointId is required")
                return
            }
            implementation.rejectConnection(endpointId: endpointId)
            call.resolve()
        } catch {
            call.reject("Reject connection failed: \(error.localizedDescription)")
        }
    }

    @objc func disconnectFromEndpoint(_ call: CAPPluginCall) {
        do {
            guard let endpointId = call.getString("endpointId") else {
                call.reject("endpointId is required")
                return
            }
            implementation.disconnectFromEndpoint(endpointId: endpointId)
            call.resolve()
        } catch {
            call.reject("Disconnect failed: \(error.localizedDescription)")
        }
    }

    @objc func disconnect(_ call: CAPPluginCall) {
        do {
            implementation.disconnect()
            call.resolve()
        } catch {
            call.reject("Disconnect failed: \(error.localizedDescription)")
        }
    }

    @objc func sendMessage(_ call: CAPPluginCall) {
        do {
            guard let endpointId = call.getString("endpointId"),
                  let data = call.getString("data") else {
                call.reject("endpointId and data are required")
                return
            }
            implementation.sendMessage(endpointId: endpointId, data: data)
            call.resolve()
        } catch {
            call.reject("Send message failed: \(error.localizedDescription)")
        }
    }

    @objc func sendFile(_ call: CAPPluginCall) {
        do {
            guard let endpointId = call.getString("endpointId"),
                  let filePath = call.getString("filePath"),
                  let fileName = call.getString("fileName") else {
                call.reject("endpointId, filePath, and fileName are required")
                return
            }
            implementation.sendFile(endpointId: endpointId, filePath: filePath, fileName: fileName)
            call.resolve()
        } catch {
            call.reject("Send file failed: \(error.localizedDescription)")
        }
    }

    @objc func setLogLevel(_ call: CAPPluginCall) {
        do {
            call.resolve()
        } catch {
            call.reject("Set log level failed: \(error.localizedDescription)")
        }
    }

    @objc func getState(_ call: CAPPluginCall) {
        do {
            call.resolve(buildStateSnapshot())
        } catch {
            call.reject("Get state failed: \(error.localizedDescription)")
        }
    }

    @objc func syncFromPlugin(_ call: CAPPluginCall) {
        do {
            let snapshot = buildStateSnapshot()
            call.resolve(snapshot)
        } catch {
            call.reject("Sync from plugin failed: \(error.localizedDescription)")
        }
    }

    private func buildStateSnapshot() -> [String: Any] {
        let endpoints = implementation.getDiscoveredEndpoints()
        let connectedEndpoints = implementation.getConnectedEndpoints()
        return [
            "endpoints": endpoints,
            "connectedEndpoints": connectedEndpoints,
            "activeTransfers": [String: Any](),
            "transferHistory": [[String: Any]](),
            "stats": [
                "totalBytesTransferred": 0,
                "filesTransferred": 0,
                "sessionStart": sessionStartTime,
                "currentSpeedBps": 0
            ]
        ]
    }

    public func onConnectionRequested(endpointId: String, endpointName: String, authToken: String) {
        notifyListeners("connectionRequested", data: [
            "endpointId": endpointId,
            "endpointName": endpointName,
            "authenticationToken": authToken,
            "isIncomingConnection": true
        ])
    }

    public func onConnectionResult(endpointId: String, status: String) {
        notifyListeners("connectionResult", data: [
            "endpointId": endpointId,
            "status": status
        ])
    }

    public func onEndpointFound(endpointId: String, endpointName: String, serviceId: String) {
        notifyListeners("endpointFound", data: [
            "endpointId": endpointId,
            "endpointName": endpointName,
            "serviceId": serviceId
        ])
    }

    public func onEndpointLost(endpointId: String) {
        notifyListeners("endpointLost", data: [
            "endpointId": endpointId
        ])
    }

    public func onMessageReceived(endpointId: String, data: String) {
        notifyListeners("messageReceived", data: [
            "endpointId": endpointId,
            "data": data
        ])
    }

    public func onTransferProgress(endpointId: String, payloadId: String, bytesTransferred: Int64, totalBytes: Int64, status: String) {
        notifyListeners("transferProgress", data: [
            "endpointId": endpointId,
            "payloadId": payloadId,
            "bytesTransferred": bytesTransferred,
            "totalBytes": totalBytes,
            "status": status
        ])
    }

    public func onFileReceived(endpointId: String, payloadId: String, fileName: String, path: String) {
        notifyListeners("fileReceived", data: [
            "endpointId": endpointId,
            "payloadId": payloadId,
            "fileName": fileName,
            "path": path
        ])
    }
}
